packages/ui-react-utils/src/safeCloneElement.ts

```diff
@@ -22,11 +22,27 @@
  * SOFTWARE.
  */
 
-import React, { ClassAttributes, ReactElement, ReactNode } from 'react'
+import React, {
+  Attributes,
+  FunctionComponentElement,
+  ComponentElement,
+  ReactElement,
+  ReactNode,
+  DOMElement
+} from 'react'
 
 import { logWarn as warn } from '@instructure/console'
 import { createChainedFunction } from '@instructure/ui-utils'
 
+type GetProps<E extends ReactElement = ReactElement> =
+  E extends FunctionComponentElement<infer P>
+    ? P & { ref?: E['ref'] }
+    : E extends ComponentElement<infer P, any>
+    ? P & { ref?: E['ref'] }
+    : E extends DOMElement<infer P, any>
+    ? P & { ref?: E['ref'] }
+    : Record<string, any> & { ref?: any }
+
```

Comment:
I haven't tried it out (and this is just nit picking really) but can't we simplify this type expression to something like this?:

```ts
type GetProps<E extends ReactElement = ReactElement> =
  E extends
  | FunctionComponentElement<infer P>
  | ComponentElement<infer P, any>
  | DOMElement<infer P, any>
  ? P & { ref?: E['ref'] }
  : Record<string, any> & { ref?: any }
```

Reply:
Yeah, I remember experimenting with sth like this but didn't work for some reason... now I tried it and worked fine :D Se I changed it, thanks!